### PR TITLE
FOUNDATIONS: Echo Task

### LINKED ISSUE #11: FOUNDATIONS: Echo Task

### PR DESCRIPTION
**Pull request by ShriHumrudha**
Closes #11

### CODE REVIEW

File: GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Validations.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            this.outputBrokerMock.Verify(broker =>
+                broker.GenerateFileOutput(It.IsAny<string>(), It.IsAny<string>()),
+                    Times.Never);
+
+            this.yamlBrokerMock.VerifyNoOtherCalls();
+            this.outputBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public void ShouldThrowValidationExceptionOnCreateIfWorkflowPathIsNull()
```
Comment (ShriHumrudha) [inline GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Validations.cs:51]: Remove Path word.

File: GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Logic.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        [Fact]
+        public void ShouldCreateWorkflow()
+        {
+            // given
+            string randomDestinationPath = GetRandomDestinationPath();
+            string inputDestinationPath = randomDestinationPath;
+            Workflow randomWorkflow = CreateRandomWorkflow();
+            Workflow inputWorkflow = randomWorkflow;
+            string serialziedWorkflow = GetRandomString();
```
Comment (hassanhabib) [inline GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Logic.cs:21]: How do we ensure this string is not going to be tampered with in your function?
for example, go into your function and do this:
serialziedWorkFlow = "BLAH!";

see if your test fails.

Comment (ShriHumrudha) [reply GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Logic.cs:21]: Yes it fails.

File: GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Exceptions.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            // given
+            Workflow someWorkflow = CreateRandomWorkflow();
+            string somePath = GetRandomDestinationPath();
+            string someMessage = GetRandomString();
+            var exception = new Exception(someMessage);
+
+            this.yamlBrokerMock.Setup(broker =>
+                broker.SerializeToYaml(It.IsAny<object>()))
+                    .Throws(exception);
+
```
Comment (hassanhabib) [inline GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Exceptions.cs:106]: Where's the localization and categorization here.

File: GitActionSharp/Models/Jobs/Machines/Machine.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public const string UbuntuLatest = "ubuntu-latest";
+        public const string Ubuntu2004 = "ubuntu-20.04";
+        public const string Ubuntu1804 = "ubuntu-18.04";
+
+        public const string WindowsLatest = "windows-Latest";
+        public const string Windows2022 = "windows-2022";
+        public const string Windows2019 = "windows-2019";
+        public const string Windows2016 = "windows-2016";
+
```
Comment (hassanhabib) [inline GitActionSharp/Models/Jobs/Machines/Machine.cs:17]: Do you need this empty line and the one above `12`?

File: GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Exceptions.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                this.workflowService.CreateWorkflow(
+                    somePath,
+                    someWorkflow);
+
+            // then
+            WorkflowServiceException actualworkflowServiceException =
+                Assert.Throws<WorkflowServiceException>(
+                    createWorkflowAction);
+
+            actualworkflowServiceException.InnerException.Message.Should()
```
Comment (hassanhabib) [inline GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Exceptions.cs:118]: We should probably add logging here.

File: GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Exceptions.cs (hunk 3 above)
Comment (hassanhabib) [reply GitActionSharp.Tests.Unit/Services/WorkflowServiceTests.Exceptions.cs:106]: We need a localized `FailedWorkflowServiceException` right? in here you passed in the native exception to your categorical service.

**Pull request merged**